File: java/app-encryption/src/main/java/com/godaddy/asherah/appencryption/envelope/EnvelopeEncryptionBytesImpl.java
package com.godaddy.asherah.appencryption.envelope;

import org.json.JSONObject;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.godaddy.asherah.appencryption.utils.Json;

public class EnvelopeEncryptionBytesImpl implements EnvelopeEncryption<byte[]> {
  private static final Logger logger = LoggerFactory.getLogger(EnvelopeEncryptionBytesImpl.class);

  private final EnvelopeEncryption<JSONObject> envelopeEncryptionJson;

  /**
   * Creates a new {@code EnvelopeEncryptionBytesImpl} instance using the provided parameters. This is an
   * implementation of {@link EnvelopeEncryption} which uses byte[] as the Data Row Record format.
   *
   * @param envelopeEncryptionJson An {@link EnvelopeEncryption} object which uses {@link org.json.JSONObject} as the
   *                               Data Row Record format.
   */
  public EnvelopeEncryptionBytesImpl(final EnvelopeEncryption<JSONObject> envelopeEncryptionJson) {
    this.envelopeEncryptionJson = envelopeEncryptionJson;
  }

  @Override
  public byte[] decryptDataRowRecord(final byte[] dataRowRecord) {
    Json dataRowRecordJson = new Json(dataRowRecord);
    return envelopeEncryptionJson.decryptDataRowRecord(dataRowRecordJson.toJsonObject());
  }

  @Override
  public byte[] encryptPayload(final byte[] payload) {
    Json drrJson = new Json(envelopeEncryptionJson.encryptPayload(payload));
    return drrJson.toUtf8();
  }

  @Override
  public void close() {
    try {
      envelopeEncryptionJson.close();
    }
    catch (Exception e) {
      logger.error("unexpected exception during close", e);
    }
  }
}

File: csharp/AppEncryption/AppEncryption/Envelope/EnvelopeEncryptionBytesImpl.cs
using System;
using System.Runtime.CompilerServices;
using GoDaddy.Asherah.AppEncryption.Util;
using Microsoft.Extensions.Logging;
using Newtonsoft.Json.Linq;

[assembly: InternalsVisibleTo("AppEncryption.Tests")]
[assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Design", "CA1816:Call GC.SuppressFinalize correctly", Justification = "This class does not have a finalizer and does not need to suppress finalization.")]

namespace GoDaddy.Asherah.AppEncryption.Envelope
{
    /// <inheritdoc />
    public class EnvelopeEncryptionBytesImpl : IEnvelopeEncryption<byte[]>
    {
        private readonly ILogger _logger;
        private readonly IEnvelopeEncryption<JObject> envelopeEncryptionJson;

        /// <summary>
        /// Initializes a new instance of the <see cref="EnvelopeEncryptionBytesImpl"/> class using the provided
        /// parameters and logger. This is an implementation of <see cref="IEnvelopeEncryption{TD}"/> which uses byte[] as the Data
        /// Row Record format.
        /// </summary>
        ///
        /// <param name="envelopeEncryptionJson">An <see cref="IEnvelopeEncryption{TD}"/> object which uses
        /// <see cref="JObject"/> as Data Row Record format.</param>
        /// <param name="logger">The logger implementation to use.</param>
        public EnvelopeEncryptionBytesImpl(IEnvelopeEncryption<JObject> envelopeEncryptionJson, ILogger logger)
        {
            this.envelopeEncryptionJson = envelopeEncryptionJson;
            this._logger = logger;
        }

        /// <inheritdoc/>
        public void Dispose()
        {
            try
            {
                envelopeEncryptionJson.Dispose();
            }
            catch (Exception e)
            {
                _logger?.LogError(e, "Unexpected exception during dispose");
            }
        }

        /// <inheritdoc/>
        public virtual byte[] DecryptDataRowRecord(byte[] dataRowRecord)
        {
            Json dataRowRecordJson = new Json(dataRowRecord);
            return envelopeEncryptionJson.DecryptDataRowRecord(dataRowRecordJson.ToJObject());
        }

        /// <inheritdoc/>
        public virtual byte[] EncryptPayload(byte[] payload)
        {
            Json drrJson = new Json(envelopeEncryptionJson.EncryptPayload(payload));
            return drrJson.ToUtf8();
        }
    }
}
